flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,368 @@
+		final int fixedOverheadPerCall = 100;
+		final int overheadPerId = 3;
+
+		int totalBytes = fixedOverheadPerCall;
+
+		for (int i = 0; i < ackIds.size(); i++) {
+			totalBytes += ackIds.get(i).length() + overheadPerId;
+			if (totalBytes > maxPayload) {
+				return Tuple2.of(ackIds.subList(0, i), ackIds.subList(i, ackIds.size()));
+			}
+		}
+
+		return Tuple2.of(ackIds, emptyList());
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
+		while (isRunning) {
+			try {
+				blockIfMaxMessagesToAcknowledgeLimitReached();
+
+				List<ReceivedMessage> messages = messagesFuture.get().getReceivedMessagesList();
+
+				// start the next pull while processing the current response.
+				messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
+
+				processMessage(sourceContext, messages);
+			} catch (InterruptedException | CancellationException e) {
+				isRunning = false;
+			}
+		}
+		shutdownSubscriber();
+	}
+
+	void processMessage(SourceContext<OUT> sourceContext, List<ReceivedMessage> messages) throws Exception {
+		synchronized (sourceContext.getCheckpointLock()) {
+			for (ReceivedMessage message : messages) {
+				acknowledgeOnCheckpoint.addAcknowledgeId(message.getAckId());
+
+				PubsubMessage pubsubMessage = message.getMessage();
+
+				OUT deserializedMessage = deserializationSchema.deserialize(pubsubMessage);
+				if (deserializationSchema.isEndOfStream(deserializedMessage)) {
+					cancel();
+					return;
+				}
+
+				sourceContext.collect(deserializedMessage);
+			}
+
+		}
+	}
+
+	private void blockIfMaxMessagesToAcknowledgeLimitReached() throws Exception {
+		while (maxMessagesToAcknowledge != -1 && getOutstandingMessagesToAck() > maxMessagesToAcknowledge) {
+			LOG.debug("Sleeping because there are {} messages waiting to be ack'ed but limit is {}", getOutstandingMessagesToAck(), maxMessagesToAcknowledge);
+			Thread.sleep(100);
+		}
+	}
+
+	private Integer getOutstandingMessagesToAck() {
+		return acknowledgeOnCheckpoint.numberOfOutstandingAcknowledgements();
+	}
+
+	@Override
+	public void cancel() {
+		isRunning = false;
+	}
+
+	@Override
+	public TypeInformation<OUT> getProducedType() {
+		return deserializationSchema.getProducedType();
+	}
+
+	/*
+	 * If we don't wait for the subscriber to terminate all background threads
+	 * ClassNotFoundExceptions will be thrown when Flink starts unloading classes.
+	 */
+	private void shutdownSubscriber() {
+		subscriber.shutdownNow();
+		eventLoopGroup.shutdownGracefully();
+		//Wait for the subscriber to terminate, to prevent leaking threads
+		while (!subscriber.isTerminated() || !eventLoopGroup.isTerminated()) {
+			try {
+				subscriber.awaitTermination(60, TimeUnit.SECONDS);
+				eventLoopGroup.awaitTermination(60, TimeUnit.SECONDS);
+			} catch (InterruptedException e) {
+				LOG.warn("Still waiting for subscriber to terminate.", e);
```

Comment:
Catching this interrupted exception in a while-loop will render this thread effectively unstoppable if the subscriber or eventLoopGroup never shut down.
I would recommend to leave the while loop on interruption. The threads should (ideally) still stop at some point.